open_spiel/games/2048.cc

```diff
@@ -0,0 +1,382 @@
+                         /*provides_observation_string=*/true,
+                         /*provides_observation_tensor=*/true,
+                         {{"max_game_length", GameParameter(kMaxGameLength)},
+                          {"max_score", GameParameter(kMaxScore)}}};
+
+std::shared_ptr<const Game> Factory(const GameParameters& params) {
+  return std::shared_ptr<const Game>(new TwoZeroFourEightGame(params));
+}
+
+REGISTER_SPIEL_GAME(kGameType, Factory);
+}  // namespace
+
+TwoZeroFourEightState::TwoZeroFourEightState(std::shared_ptr<const Game> game)
+    : State(game),
+      board_(std::vector<Tile>(kDefaultRows * kDefaultColumns, Tile(0, false)))
+      {}
+
+void TwoZeroFourEightState::SetCustomBoard(const std::vector<int>& board_seq) {
+  current_player_ = 0;
+  for (int r = 0; r < kDefaultRows; r++) {
+    for (int c = 0; c < kDefaultColumns; c++) {
+      SetBoard(r, c, Tile(board_seq[r * kDefaultRows + c], false));
+    }
+  }
+}
+
+ChanceAction TwoZeroFourEightState
+    ::SpielActionToChanceAction(Action action) const {
+  std::vector<int> values = UnrankActionMixedBase(
+      action, {kDefaultRows, kDefaultColumns, kChanceTiles.size()});
+  return ChanceAction(values[0], values[1], values[2]);
+}
+
+Action TwoZeroFourEightState
+    ::ChanceActionToSpielAction(ChanceAction move) const {
+  std::vector<int> action_bases = {kDefaultRows, kDefaultColumns, 
+      kChanceTiles.size()};
+  return RankActionMixedBase(
+      action_bases, {move.row, move.column, move.is_four});
+}
+
+std::vector<std::vector<int>> TwoZeroFourEightState
+    ::BuildTraversals(int direction) const {
+  std::vector<int> x, y;
+  for (int pos = 0; pos < kDefaultRows; pos++) {
+    x.push_back(pos);    
+  }
+  for (int pos = 0; pos < kDefaultColumns; pos++) {
+    y.push_back(pos);    
+  }
+  switch (direction) {
+    case kMoveRight:
+      reverse(x.begin(), x.end());
+      reverse(y.begin(), y.end());
+      break;
+    case kMoveLeft:
+    case kMoveDown:
+      reverse(x.begin(), x.end());
+      break;
+  }
+  return {x, y};
+};
+
+bool TwoZeroFourEightState::WithinBounds(int r, int c) const {
+  return r >= 0 && r < kDefaultRows && c >= 0 && c < kDefaultColumns;
+};
+
+bool TwoZeroFourEightState::CellAvailable(int r, int c) const {
+  return BoardAt(r, c).value == 0;
+}
+
+Coordinate GetVector(int direction) {
+  switch (direction) {
+      case kMoveUp:
+        return Coordinate(-1, 0);        
+      case kMoveRight:
+        return Coordinate(0, 1);
+      case kMoveDown:
+        return Coordinate(1, 0);
+      case kMoveLeft:
+        return Coordinate(0, -1);
+    }
+}
+
+std::vector<Coordinate> TwoZeroFourEightState
+    ::FindFarthestPosition(int r, int c, int direction) const {  
+  // Progress towards the vector direction until an obstacle is found
+  Coordinate prev = Coordinate(r, c);
+  do {
+    prev = Coordinate(r, c);
+    Coordinate direction_diff = GetVector(direction);
```

Comment:
Given the direction doesn't change, the vector is constant. The compiler might figure that out, but you could force it by hoisting it out of the loop. Making it constexpr would also do it.

Reply:
Changed this